src/Analyser/TypeSpecifier.php

```diff
@@ -820,47 +821,8 @@ public function specifyTypesInCondition(
 						);
 					} else {
 						$varType = $scope->getType($var->var);
-						if ($varType->isArray()->yes() && !$varType->isIterableAtLeastOnce()->no()) {
-							$varIterableKeyType = $varType->getIterableKeyType();
-
-							if ($varIterableKeyType->isConstantScalarValue()->yes()) {
-								$narrowedKey = TypeCombinator::union(
-									$varIterableKeyType,
-									TypeCombinator::remove($varIterableKeyType->toString(), new ConstantStringType('')),
-								);
-
-								if (!$varType->hasOffsetValueType(new ConstantIntegerType(0))->no()) {
-									$narrowedKey = TypeCombinator::union(
-										$narrowedKey,
-										new ConstantBooleanType(false),
-									);
-								}
-
-								if (!$varType->hasOffsetValueType(new ConstantIntegerType(1))->no()) {
-									$narrowedKey = TypeCombinator::union(
-										$narrowedKey,
-										new ConstantBooleanType(true),
-									);
-								}
-
-								if (!$varType->hasOffsetValueType(new ConstantStringType(''))->no()) {
-									$narrowedKey = TypeCombinator::addNull($narrowedKey);
-								}
-
-								if (!$varIterableKeyType->isNumericString()->no() || !$varIterableKeyType->isInteger()->no()) {
-									$narrowedKey = TypeCombinator::union($narrowedKey, new FloatType());
-								}
-							} else {
-								$narrowedKey = new MixedType(
-									false,
-									new UnionType([
-										new ArrayType(new MixedType(), new MixedType()),
-										new ObjectWithoutClassType(),
-										new ResourceType(),
-									]),
-								);
-							}
-
+						$narrowedKey = AllowedArrayKeysTypes::narrowOffsetKeyType($varType, $dimType);
```

Comment:
as a first step I think its easier to have a static method for this.

after we get more similar with the `array_key_exists` impl we will see whether it makes sense to move this method into the `ConstantScalarType`-interface, as we discussed in dresden